File: code/cpp/findComponents.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include "helpers.h"


void dfs(networks::Graph &graph, int node, std::vector<bool> &visited, std::vector<int> &components, int count)
{
    visited[node] = true;
    components[node] = count;
    for (auto next_node : graph.adjList[node])
    {
        if (!visited[next_node.first])
            dfs(graph, next_node.first, visited, components, count);
    }
}

std::pair<int, std::vector<int>> findComponents(networks::Graph &graph, std::vector<bool> &visited, int count=0)
{
    std::vector<int> components(graph.numNodes);
    for (int i=0; i < graph.numNodes; i++)
    {
        if (!visited[i])
        {
            count++;
            dfs(graph, i, visited, components, count);
        }
    }

    return {count, components};
}


int main()
{
#ifndef FILE_INPUT
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif

    int numNodes, numEdges;
    std::cin >> numNodes >> numEdges;

    bool directed, weighted;
    std::cin >> directed >> weighted;

    std::vector<networks::Edge> edges;
    if (weighted)
        for (int i=0; i<numEdges; i++)
        {
            int u, v, w;
            std::cin >> u >> v >> w;
            edges.push_back(networks::Edge(u, v, w)); 
        }

    else
        for (int i=0; i<numEdges; i++)
        {
            int u, v;
            std::cin >> u >> v;
            edges.push_back(networks::Edge(u, v));
        }

    networks::Graph graph = networks::Graph(numNodes, edges);
    std::vector<bool> visited(numNodes, false);

    std::pair<int, std::vector<int>> ans = findComponents(graph, visited);
    std::cout << "No. of connected components: " << ans.first;
    helpers::printVector(ans.second, "The Components array is: ");
}

File: code/python/find_components.py
import helpers as H

# global variables
INPUT_PATH = 'input.txt'
OUTPUT_PATH = 'output.txt'

# clearing the contents of the file if any
open(OUTPUT_PATH, 'w').close()


def dfs(graph: H.Graph, node: int, visited: list, componets: list, count: int):
    visited[node] = True
    componets[node] = count
    
    for next_node in graph.adj_list[node]:
        if not visited[next_node[0]]:
            dfs(graph, next_node[0], visited, componets, count)
            
            
def find_components(graph: H.Graph, visited: list, count=0):
    components = [None for _ in range(graph.num_nodes)]
    for i in range(graph.num_nodes):
        if not visited[i]:
            count += 1
            dfs(graph, i, visited, components, count)
            
    return count, components


if __name__ == '__main__':
    with open(INPUT_PATH, 'r') as f:
        num_nodes, num_edges = (int(i) for i in f.readline().split(" "))
        directed, weighted = (int(i) for i in f.readline().split(" "))
        
        edges = []
        if weighted:
            for i in range(num_edges):
                u, v, w = (int(i) for i in f.readline().split(" "))
                edges.append(H.Edge(u, v, w))
        else:
            for i in range(num_edges):
                u, v = (int(i) for i in f.readline().split(" "))
                edges.append(H.Edge(u, v))
                
    graph = H.Graph(num_nodes, edges, directed)
    visited = [False for _ in range(num_nodes)]
        
    ans = find_components(graph, visited)
    with open(OUTPUT_PATH, 'a') as f:
        f.write(f"Number of Components: {ans[0]}\nComponents Array: {ans[1]}")